### DISCUSSION TITLE
Strange Ableton parameter control issue

### DISCUSSION THREAD

**Discussion by puffer3** [RomanKubiak/ctrlr#463] [Q&A]
When enabling parameters for ableton control they have a substantial amount of decimal places that seem to be bogging down my CPU.  Has anyone else experienced this issue?  It makes automating controls on the panel a major drain on the computer when it should really be no problem.  

Is there any way to limit the value of the control range for parameters?  I am not sure exactly how to fix this issue. 

I mentioned this on my AU thread but this is sort of its own problem so I am starting another thread here. 

![Control_values](https://user-images.githubusercontent.com/6006231/168147400-9a6d75ca-a1b2-4229-be27-5549f1e96237.jpg)

**Comment by dnaldoog** [comment]
What are the VST numbers? Are they very large? Sometimes Ctrlr seems to generate large VST indices. Maybe that's affecting things? But, I really have no idea, but if it were me I would try regenerating them starting at `1`.

Here is a script for that and example panel:

[Prophet 600 gligli (with unblock button)_0.0_12_VST INDEX GENERATOR.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8682973/Prophet.600.gligli.with.unblock.button._0.0_12_VST.INDEX.GENERATOR.panel.zip)


<pre>
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
generateVstIndex = function( mod, value , source)
    if myPanel:getBootstrapState() then
        return
    end
    if source ~= 4 then
        return
    end
    local _toBeVstIndexed = {}

    for k1 in pairs(patchDataTable) do
        table.insert(_toBeVstIndexed, k1)
    end
    --[[
local extra={"ccmappingnumber","chooseSnapshot","snapShotSelect"} -- add other modulator names here

for i,v in ipairs(extra) do
table.insert(_toBeVstIndexed,v)

end
--]]
    local no = panel:getNumModulators()
    for i = 0, no - 1 do
        --CLEAR ALL FIRST
        mod = panel:getModulatorByIndex(i)
        mod:setPropertyString("vstIndex", "")
        mod:setPropertyString("modulatorVstExported", "0")
    end
    for i, v in ipairs(_toBeVstIndexed) do
        local n = panel:getModulatorByName(v)
        n:setPropertyString("vstIndex", tostring(i))
        n:setPropertyString("modulatorVstExported", "1")
        console(String(string.format('"%s", --[[ set to %d--]]', v, i)))
    end
    console(String("table size=" .. #_toBeVstIndexed))
end --function
</pre>

**Reply by damiensellier** [reply]
Yeah exactly, not sure if indexing the vst parameter from 1 will help but Ctrlr is very strange about assigning this parameter with high numbers.
In the source code of Ctrlr, there's a file specific to Ableton live "[CtrlrProcessorEditorForLive.cpp](https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/Plugin/CtrlrProcessorEditorForLive.cpp)" that sets up the visible portion of the panel inside the host. There's this old 24px offset trick for the menu in it.  So there's probably other Live-specific code somewhere else that messes up the menu and probably how it handles the parameter indexes. I never used Ableton Live so I can't tell precisely the caveats with Ctrlr.

**Reply by puffer3** [reply]
I renumbered my VST values to be lower.  But even then the values in that screenshot have to do with the range of the controls.  Its computing what ends up being a 0-127 midi message to 6 decimal places.  Its a resource drag that makes automation not viable.  I cant figure out if theres some way to reduce the scale of the control. 


I am having another issue where at the beginning of every clip on the timeline it is sending a snapshot to the synth.  this is also something I need to prevent happening automatically if I want to actually use the panels to make music heheh   This was happening in fruity loops too.  I don't think it has to do with "send snapshot on program change" or anything like that.  Has anyone else experienced this with their daw?

**Comment by puffer3** [comment]
I know this isnt quite relevant to dev but it was important for me to figure out:

okay so as for the strange automation data that was being sent on a per clip basis in the timeline of ableton-

there are "clip envelopes" that are MIDI data recorded into the clip/midi data.  This doesnt show up in the regular automation lanes, so it just seems like weird mystery midi CC data on the midi monitor.  It took me a long time to find it and then copy the notes to a new clip and leave all the unwanted "envelopes" or midi control data behind (where did it even come from??)  

still trying to figure out the variable issue.  I also had a similar CC problem in FL studio but I will have to track that down some other way.

**Comment by dobo365** [comment]
When a CC is supposed to be 0-127 it is only integer data.
Is it your own panel? What do you see in the Midi monitor output?
The value shown in Ableton is maybe some normalized 0-1 visualization "value/127"
Did you put something in the property "Expression to evaluate when calculating the Midi message value from the modulator value"?

**Reply by damiensellier** [reply]
I think it's problably coming from Ableton itself. if ever you have a problem with those values you can replace your sliders with "uiFixedSlider" instead.

![replace by uiFixedSlider](https://user-images.githubusercontent.com/55164385/169675446-07fe8640-05f0-4232-a99e-36fc418e08e1.png)




 with **"Slider contents"** 0 to 127. Since it will be fixed, you won't get those decimals, unless it's on ableton's side.

![uiFixedSlider 0-127](https://user-images.githubusercontent.com/55164385/169675138-672bf2e5-94c6-4a83-b864-a1e939b7a53a.png)


From those fixed values, you can use :
ModValToInt = mod:getModulatorValue() -- Get Modulator Value
--console("Displays ModValToInt : "..(ModValToInt))

Let me know if it fixed the issue or it is still showing decimals in ableton ;)

**Reply by puffer3** [reply]
I have some fixed sliders in my panel for oscillator tuning (so I can display notes instead of numbers for oscillator values) and the decimals are still there.  I think this has to do more with the way ctrlr and ableton are talking.   

here's my current OBsX Panel if that helps.
[OB-SX_1_3_OBsXdiskSaving_2022-05-22_01-40.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/8749071/OB-SX_1_3_OBsXdiskSaving_2022-05-22_01-40.bpanelz.zip)

I went and looked at how something like Zebra2 softsynth looks in this mode and the results were parameter specific, but terminated at two decimal points.  There must be a way to set the order of magnitude of the variables.
![Screen Shot 2022-05-22 at 1 42 54 AM](https://user-images.githubusercontent.com/6006231/169686344-ca856ab9-f1f3-4d15-a843-5936e1a37611.png)

**Reply by damiensellier** [reply]
It really goes beyond my knowledge, I cannot help you about the ableton live problem, I'm sorry Henry.

**Reply by Godlike-Productions** [reply]
With ableton, you can set the sliders to MIDI CC mode (0-127), which sill probably work better.
